examples/step-69/step-69.cc

```diff
@@ -0,0 +1,2926 @@
+                   const std::string &subsection = "Discretization");
+
+    void setup();
+
+    const MPI_Comm &mpi_communicator;
+
+    parallel::distributed::Triangulation<dim> triangulation;
+
+    const MappingQ<dim>   mapping;
+    const FE_Q<dim>       finite_element;
+    const QGauss<dim>     quadrature;
+    const QGauss<dim - 1> face_quadrature;
+
+  private:
+    TimerOutput &computing_timer;
+
+    double length;
+    double height;
+    double disc_position;
+    double disc_diameter;
+
+    unsigned int refinement;
+  };
+
+  // @sect4{The <code>OfflineData</code> class}
+  //
+  // The class <code>OfflineData</code> contains pretty much all components
+  // of the discretization that do not evolve in time, in particular, the
+  // DoFHandler, SparsityPattern, boundary maps, the lumped mass,
+  // $\mathbf{c}_{ij}$ and $\mathbf{n}_{ij}$ matrices. Here, the term
+  // <i>offline</i> refers to the fact that all the class
+  // members of <code>OfflineData</code> have well-defined values
+  // independent of the current time step. This means that they can be
+  // initialized ahead of time (at <i>time step zero</i>) and are not meant
+  // to be modified at any other later time step. For instance, the
+  // sparsity pattern should not change as we advance in time (we are not
+  // doing any form of adaptivity in space). Similarly, the entries of the
+  // lumped mass matrix should not be modified as we advance in time
+  // either.
+  //
+  // We also compute and store a <code>boundary_normal_map</code> that
+  // contains a map from a global index of type `types:global_dof_index` of
+  // a boundary degree of freedom to a tuple consisting of a normal vector,
+  // the boundary id, and the position associated with the degree of
+  // freedom. We actually have to compute and store this geometric
+  // information in this class because we won't have access to geometric
+  // (or cell-based) information later on in the algebraic loops over the
+  // sparsity pattern.
+  //
+  // @note Even though this class currently does not have any parameters
+  // that could be read in from a parameter file we nevertheless derive
+  // from ParameterAcceptor and follow the same idiom of providing a
+  // <code>setup()</code> (and <code>assemble()</code>) method as for the
+  // class Discretization.
+
+  template <int dim>
+  class OfflineData : public ParameterAcceptor
+  {
+  public:
+    using BoundaryNormalMap =
+      std::map<types::global_dof_index,
+               std::tuple<Tensor<1, dim>, types::boundary_id, Point<dim>>>;
+
+    OfflineData(const MPI_Comm &           mpi_communicator,
+                TimerOutput &              computing_timer,
+                const Discretization<dim> &discretization,
+                const std::string &        subsection = "OfflineData");
+
+    void setup();
+    void assemble();
+
+    DoFHandler<dim> dof_handler;
+
+    std::shared_ptr<const Utilities::MPI::Partitioner> partitioner;
```

Comment:
Is it really shared, or could it be a `std::unique_ptr`?

Reply:
It is really shared (the `partitioner` is stored as a `shared_ptr` inside the distributed vector class, and the constructor requires to pass a shared pointer).